File: Go/Go_1000-/1089.go
// Go
// leetcode 1089
// https://leetcode.cn/problems/duplicate-zeros/

func duplicateZeros(arr []int)  {
	// 统计
	cnt, diff, n := 0, 0, len(arr)
	for i := 0; i + cnt < n; i++ {
		if arr[i] == 0 {
			cnt++
		}
		if i + cnt == n {
			diff = 1
		}
	}
	// 预防类似 [1, ..., 1, 0] 这种情况
	if(diff == 1) {
		cnt--
		arr[n - 1] = 0
	}

	for i, j := n-1-diff, n-1-cnt-diff; j >= 0 && i != j; i-- {
		arr[i] = arr[j]
		if(arr[j] == 0) {
			i--
			arr[i] = 0
		}
        j--
	}
}

File: cpp/cpp_1000-/1089.cc
// C++
// leetcode 1089
// https://leetcode.cn/problems/duplicate-zeros/

class Solution {
public:
    void duplicateZeros(vector<int>& arr) {
    	int cnt = 0, n = arr.size(), diff = 0;
    	for(int i = 0; i + cnt < n; i++) {
    		if(arr[i] == 0)
    			cnt++;
            diff = max(i + cnt - n + 1, 0);
    	}

    	// diff 预防的最典型例子
    	// 很多非0数字 + 末尾一个0
    	if(diff == 1) {
    		arr[n-1] = 0;
            cnt--;
    	}

    	// 填充
    	for(int i = n-1-diff, j = n-1-cnt-diff; j >= 0 && i != j; i--, j--) {
    		arr[i] = arr[j];
    		if(arr[j] == 0) {
    			arr[--i] = 0;
    		}
    	}
    }
};